zeroknowledgeproof/src/main/java/com/zeroknowledgeproof/rangeProof/RangeProofProver.kt

```diff
@@ -0,0 +1,199 @@
+        /// Step 0: Create initial commitment
+        val r = generateRandomInterval(ZERO, k2)
+        val c = g.modPow(toBigInt(m), N).times(h.modPow(r, N)).mod(N)
+
+        /// Step 1: calculate c1 and c2
+        val c1 = (c * calculateInverse(g.pow(a - 1), N)).mod(N)
+        val c2 = (g.pow(b + 1) * calculateInverse(c, N)).mod(N)
+        println(">STEP 1 FINISHED")
+
+        // Step 2: Generate rPrime, calculate cPrime, publish it.
+        val rPrime = generateRandomInterval(ZERO, k2)
+        val cPrime = (c1.modPow(toBigInt(b - m + 1), N) * h.modPow(rPrime, N)).mod(N)
+
+        val proofEqualIntegers = proveTwoCommittedIntegersAreEqual(toBigInt(b - m + 1), -r, rPrime, g, h, c1, h, c2, cPrime)
+        println(">STEP 2 FINISHED")
+
+        // Step 3: Choose w from Zk2 - {0}, and r'' from Zk2
+        val w = generateRandomInterval(ONE, k2) // Lower bound is 1
+        val rDPrime = generateRandomInterval(ZERO, k2)
+
+        // Publicly give a proof that two integers are equal
+        val cDPrime = (cPrime.modPow(w * w, N) * h.modPow(rDPrime, N)).mod(N) // The commitment of w in base cPrime
+        val cDPrimeIsSquare = proveCommittedNumberIsSquare(w, rDPrime, cPrime, h, cDPrime)
+        println(">STEP 3 FINISHED")
+
+        // Step 4: Generate m1, m2, m3
+        val sum = w * w * toBigInt((m - a + 1) * (b - m + 1))
+        val (m1, m2, m4) = calculateMValues(sum)
+        val m3 = m4 * m4
+
+        // Generate r1, r2, r3 such that their sum is equal to w^2 (( b - m + 1)r + r') + r''
+        val (r1, r2, r3) = calculateRValues(w * w * (toBigInt(b - m + 1) * r + rPrime) + rDPrime)
+        // Then, prove that m3 is a square.
+        val c1Prime = (g.modPow(m1, N) * h.modPow(r1, N)).mod(N)
+        val c2Prime = (g.modPow(m2, N) * h.modPow(r2, N)).mod(N)
+        val c3Prime = (cDPrime * calculateInverse(c1Prime * c2Prime, N)).mod(N)
+        val m3IsSquare = proveCommittedNumberIsSquare(m4, r3, g, h, c3Prime)
+
+        println(">STEP 4 FINISHED")
+
+        // Put all results in the result struct, return it.
+        val pubRes = SetupPublicResult(g = g, h = h, c = c, c1 = c1, c2 = c2, cPrime = cPrime,
+                cDPrime = cDPrime, c1Prime = c1Prime, c2Prime = c2Prime, c3Prime = c3Prime,
+                sameCommitment = proofEqualIntegers, cDPrimeIsSquare = cDPrimeIsSquare,
+                m3IsSquare = m3IsSquare, k1 = BigInteger(2048, rand))
+        // To perform the verification, the verifier needs a bound. This bound is k1, generated by the prover.
+
+        // Save the private parameters, that the verifier should NOT obtain
+        val privRes = SetupPrivateResult(m1 = m1, m2 = m2, m3 = m3, r1 = r1, r2 = r2, r3 = r3)
+        return Pair(pubRes, privRes)
+    }
+
+    fun answerUniqueChallenge(privSetup: SetupPrivateResult, challenge: Challenge): InteractivePublicResult {
+        // Take values from relative structs
+        // private values
+        val (m1, m2, m3, r1, r2, r3) = privSetup
```

Comment:
If we are destructuring every value of privSetup here anyways to bring them all in scope, why not make this a method on SetupPrivateResult